network/wsNetwork.go

```diff
@@ -2148,6 +2205,38 @@ func (wn *WebsocketNetwork) tryConnect(addr, gossipAddr string) {
 		return
 	}
 
+	var peerPublicKey crypto.PublicKey
+	var idChalResp identityChallengeResponse
+	if shouldSupportIdentityChallenge(matchingVersion) {
+		idChalRespB64 := response.Header.Get(ProtocolConectionIdentityChallengeHeader)
+		if idChalRespB64 == "" {
+			wn.log.Warnf("ws connect(%s) no identification challenge", gossipAddr)
+			return
+		}
+		idChalResp = IdentityChallengeResponseFromB64(idChalRespB64)
+		// if the identification challenge response is not correctly signed, do not proceed
+		if err = idChalResp.verify(); err != nil {
+			wn.log.Warnf("ws connect(%s) failed identification challenge signature: %s", gossipAddr, err.Error())
+			return
+		}
+		// if the signed challenge does not return our original challenge, do not proceed
+		if idChalResp.Challenge != idChallenge {
+			wn.log.Warnf("ws connect(%s) failed identification challenge match: %s", gossipAddr, idChallenge)
+			return
+		}
+		// now that we have validated the identity challenge, we can proceed and treat this peer as validated
+		peerPublicKey = idChalResp.Key
+		// if we already have a verified peer using this key, do not proceed
+		wn.peersLock.RLock()
+		existingPeer := wn.peersByID[peerPublicKey]
+		wn.peersLock.RUnlock()
+		if existingPeer != nil &&
```

Comment:
Here, the node is trying to connect to a peer already connected with another connection (call connection X).
But the other peer has already accepted this connection request (call connection Y), which means, the other peer will reject the connection X, and this node will reject connection Y.

Can this situation happen? where both connections fail, and the two peers are not connected.

Reply:
It could happen if the peer on the other side is behaving maliciously. Since both sides need to confirm each-other's identities, they both have opportunities to disconnect

Reply:
I'm not fully convinced yet. I tried to add some more detail in a comment on `wn.addPeer(peer)`